Add SGX Enclave  Memory Manager

haitaohuang opened the pull request:
Signed-off-by: Haitao Huang <4699115+haitaohuang@users.noreply.github.com>

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+
+## Motivation ##
+
+An enclave's memory is backed by special reserved region in RAM, called Enclave Page Cache(EPC). Enclave memory management tasks include allocating/reserving virtual address ranges, committing physical EPC pages, changing EPC page permissions or page type, or removing EPC pages. Those tasks require collaboration between trusted runtime, untrusted runtime and the OS. The SGX enclave memory manager ...
```
nit: please split lines into (say) 80 column width, which makes git diffs easier to follow.

haitaohuang replied:
will do for next version

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+
+## Motivation ##
+
+An enclave's memory is backed by special reserved region in RAM, called Enclave Page Cache(EPC). Enclave memory management tasks include allocating/reserving virtual address ranges, committing physical EPC pages, changing EPC page permissions or page type, or removing EPC pages. Those tasks require collaboration between trusted runtime, untrusted runtime and the OS. The SGX enclave memory manager ...
```
```suggestion
An enclave's memory is backed by a special reserved region in RAM, called Enclave Page Cache (EPC). Enclave memory management tasks include allocating/reserving virtual address ranges, committing physical EPC pages, changing EPC page permissions or page type, and removing EPC pages. Those tasks require collaboration between the trusted runtime, the untrusted runtime, and the OS. The SGX enclave memory manager (EMM) serves as a central component in the enclave trusted runtime that abstracts the interaction with the untrusted runtime for all memory management flows and provides APIs for its clients to reserve virtual address ranges, commit EPC memory to the reserved address ranges, and modify attributes of the reserved/committed pages.   
```
fix grammar

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+
+The clients can also call the EMM commit API, sgx_mm_commit, to proactively commit specific sub-regions in a COMMIT_ON_DEMAND allocation to avoid future page fault.
+
+Some EMM clients, <i>e.g.</i>, a dynamic code loader wishing to load code on page fault, can register a custom handler for page fault at the time of allocation request. In the custom page fault handler, it can invoke an API, sgx_mm_commit_data, to commit and load data to newly committed EPC page at the same time as supported by EACCEPTCOPY. 
```
```suggestion
Some EMM clients, <i>e.g.</i>, a dynamic code loader wishing to load code on page faults, can register a custom handler for page faults at the time of allocation request. In the custom page fault handler, it can invoke an API, sgx_mm_commit_data, to commit and load data to newly committed EPC page at the same time as supported by EACCEPTCOPY. 
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+
+## Notes on Internal Design ##
+
+The enclave memory manager keeps track of memory allocation and layout info inside enclave address range (ELRANGE) using an internal structure called Enclave Memory Area (EMA) List. The EMA and EMA list are considered private data structure of memory manager, and their internals are not exposed in client-facing APIs.
```
```suggestion
The enclave memory manager keeps track of memory allocation and layout info inside enclave address range (ELRANGE) using an internal structure called the Enclave Memory Area (EMA) List. The EMA and EMA list are considered private data structure of memory manager, and their internals are not exposed in client-facing APIs.
```

Also I cannot parse the grammar of "EMA and EMA list are considered private data structure of memory manager".   structure is singular but 'are' is plural.  Are "EMA" and "EMA list" two different things?

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+**Assumptions:**
+
+- When enclave is loaded, the OS has reserved the whole address range covered by ELRANGE. It is assumed the host app will not remap any part of this reserved range.
+  - In future, when the new feature that allows loading enclave with base address at zero, we may support partial ELRANGE reserved by the OS.  
```
```suggestion
  - In the future, when the new feature that allows loading an enclave with base address at zero, we may support partial ELRANGE reserved by the OS.  
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+
+- When enclave is loaded, the OS has reserved the whole address range covered by ELRANGE. It is assumed the host app will not remap any part of this reserved range.
+  - In future, when the new feature that allows loading enclave with base address at zero, we may support partial ELRANGE reserved by the OS.  
+- Memory manager does not check EPC pressure, or proactively trim pages when EPC runs low. The OS can reclaim EPC pages when EPC running low or cgroups threshold reached
```
```suggestion
- The memory manager does not check EPC pressure, or proactively trim pages when EPC runs low. The OS can reclaim EPC pages when EPC is running low or the cgroups threshold is reached
```

Also regarding "when cgroups threshold reached": how does this apply to Windows?

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+- When enclave is loaded, the OS has reserved the whole address range covered by ELRANGE. It is assumed the host app will not remap any part of this reserved range.
+  - In future, when the new feature that allows loading enclave with base address at zero, we may support partial ELRANGE reserved by the OS.  
+- Memory manager does not check EPC pressure, or proactively trim pages when EPC runs low. The OS can reclaim EPC pages when EPC running low or cgroups threshold reached
+- Memory manager does not maintain and recycle committed then freed pages
```
```suggestion
- The memory manager does not maintain and recycle committed then freed pages
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+  - In future, when the new feature that allows loading enclave with base address at zero, we may support partial ELRANGE reserved by the OS.  
+- Memory manager does not check EPC pressure, or proactively trim pages when EPC runs low. The OS can reclaim EPC pages when EPC running low or cgroups threshold reached
+- Memory manager does not maintain and recycle committed then freed pages
+  - Whenever a page is freed (via dealloc or uncommit API), it is trimmed from enclave and need be re-allocated and committed before re-use.
```
```suggestion
  - Whenever a page is freed (via dealloc or uncommit API), it is trimmed from enclave and needs to be re-allocated and committed before re-use.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+- Memory manager does not check EPC pressure, or proactively trim pages when EPC runs low. The OS can reclaim EPC pages when EPC running low or cgroups threshold reached
+- Memory manager does not maintain and recycle committed then freed pages
+  - Whenever a page is freed (via dealloc or uncommit API), it is trimmed from enclave and need be re-allocated and committed before re-use.
+  - Owner of a region can re-purpose a sub-region of it by calling sgx_mm_modify_type/permissions to split out the sub-region to be reused.
```
```suggestion
  - The owner of a region can re-purpose a sub-region of it by calling sgx_mm_modify_type/permissions to split out the sub-region to be reused.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+- Memory manager does not maintain and recycle committed then freed pages
+  - Whenever a page is freed (via dealloc or uncommit API), it is trimmed from enclave and need be re-allocated and committed before re-use.
+  - Owner of a region can re-purpose a sub-region of it by calling sgx_mm_modify_type/permissions to split out the sub-region to be reused.
+- Memory manager does not call back client for #GP handling. Memory manager code will ensure itself would not cause #GP, and only register a #PF handler with the enclave global exception handler registry. Clients wish to handle #GP can register its own exception handler with the global handler registry.
```
```suggestion
- The memory manager does not call back into the client for #GP handling. Memory manager code will ensure that it itself would not cause #GP, and only registers a #PF handler with the enclave global exception handler registry. A client wanting to handle #GP can register its own exception handler with the global handler registry.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+- Memory manager does not call back client for #GP handling. Memory manager code will ensure itself would not cause #GP, and only register a #PF handler with the enclave global exception handler registry. Clients wish to handle #GP can register its own exception handler with the global handler registry.
+
+
+Public APIs
```
In which library file?

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+ * @param[in] pfinfo, info reported in the SSA MISC region for page fault
+ * @param[in] private_data, private data provided by handler in sgx_mm_alloc call.
+ * @return EXCEPTION_CONTINUE_EXECUTION on success handling the exception.
+ *    abort if internal corruption detected, Otherwise EXCEPTION_CONTINUE_SEARCH.
```
```suggestion
 *    abort if internal corruption detected. Otherwise EXCEPTION_CONTINUE_SEARCH.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+*                              are done when this function returns.
+*            - EMA_COMMIT_ON_DEMAND: reserves memory range, EPC pages are committed (EACCEPT) on demand upon #PF.
+*        ORed with zero or one of the committing order flags:
+*            - EMA_GROWSDOWN: always commit pages from higher to lower addresses, no gaps in addresses above the last commited
+*            - EMA_GROWSUP:   always commit pages from lower to higher addresses, no gaps in addresses below the last commited.
```
```suggestion
*            - EMA_GROWSDOWN: always commit pages from higher to lower addresses, no gaps in addresses above the last committed.
*            - EMA_GROWSUP:   always commit pages from lower to higher addresses, no gaps in addresses below the last committed.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+ * 
+ * @param[in] pfinfo, info reported in the SSA MISC region for page fault
+ * @param[in] private_data, private data provided by handler in sgx_mm_alloc call.
+ * @return EXCEPTION_CONTINUE_EXECUTION on success handling the exception.
```
use `retval` not `return` for discrete values

haitaohuang replied:
will do for next version

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+/**
+* Uncommit (trim) physical EPC pages in a previously committed range. 
+* The pages in the allocation are freed, but the address range is still reserved.
+* @param[in] addr page aligned start address of the region to be trimmed
+* @param[in] length size in bytes of multiples of page size.
```
nit: be consistent in terms of whether argument descriptions in in period or not

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+Public APIs
+-----------------
+
+### sgx_mm_alloc
```
Also, are these proposed as APIs to be added to OE?  (where APIs would be prefixed with `oe_`)

haitaohuang replied:
The intent is to reuse implementation of these APIs between OE and Intel SGX sdk. I'm not very familiar with OE yet. From what I see, if we decide to expose these as common APIs across different technologies for OE, then we need add these APIs with oe_ prefix, and do weak alias for SGX impl?

dthaler replied:
Do you really need a weak alias?  I suspect a simple #define macro that defines one to be an alias for the other, would be sufficient.

dthaler replied:
Also I don't see any text in this document about "reuse implementation".   Please explain the code relationship.  Does one depend on the other or should it be in a separate common library consumed by both?

haitaohuang replied:
Per discussion in arch forum, we plan to reuse it as lib

mingweishih commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+
+## Motivation ##
+
+An enclave's memory is backed by a special reserved region in RAM, called Enclave Page Cache (EPC). Enclave memory management tasks include allocating/reserving virtual address ranges, committing physical EPC pages, changing EPC page permissions or page type, and removing EPC pages. Those tasks require collaboration between the trusted runtime, the untrusted runtime, and the OS. The SGX enclave m ...
```
Few comments/questions:
- The paragraph mentions "... collaboration among trusted runtime, untrusted runtime, and the OS." However, I don't see the OS is mentioned in the following sentences.
- Does EMM run in a dedicated enclave or part of each enclave application? That is, does a "client" refer to the enclave which talks to the EMM enclave or the enclave includes the EMM. Can you clarify a bit more?

haitaohuang replied:
EMM does not directly work with OS, but rather thru interaction with  untrusted time. So no need to mention OS.

I mentioned typical clients below: The public EMM APIs defined here are most likely invoked by some intermediate runtime level components for specific usages, such as dynamic heap/stack, mmap, mprotect, higher level language JIT compiler, etc.

I think special reserved memory sounds better in this case or I should use a different term. I'll take other suggestions in next version

mingweishih commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+ * @return EXCEPTION_CONTINUE_EXECUTION on success handling the exception.
+ *    abort if internal corruption detected. Otherwise EXCEPTION_CONTINUE_SEARCH.
+ */
+typedef int (*enclave_fault_handler_t)(const sgx_pfinfo* pfinfo, void* private_data);
```
Note that if we want to implement this on OE, we should reuse OE's exception handler interfaces.

haitaohuang replied:
This is a callback invoked by EMM, speciifc for PF. EMM implementation itself would need register runtime specific exception handler interface with the global registry

mingweishih commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+        };
+    } pfec;
+    uint32_t reserved; 
+} sgx_pfinfo;
```
Note from OE: the similar data structure will be supported by OE as part of #3984.

mingweishih commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+ * @return 0 on success, or error code similar to mprotect
+ */
+
+mprotect_ocall(void*addr, int flags)
```
missing function type and `;` at the end

mingweishih commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+
+The memory manager must be initialized in first ECALL (ECMD_INIT_ENCLAVE in Intel SGX SDK) before any other clients can use it. Therefore, code and data of the memory manager will be part of initial enclave image that are loaded with EADD before EINIT, and as a part of the trusted runtime. 
+
+The trusted runtime should enumerate all initial committed regions (code, data, heap, stack, TCS, SSA), and call the EMM internal APIs to setup initial entries in the EMA list to track existing regions and mark some of them not modifiable by EMM public APIs. Runtime also ensures there is enough reserved space on heap for EMM to create the initial EMA list and the entries. Once initialized, the me ...
```
Note for OE: This option indicates some work to the OE loader. Currently we don't explicitly pass the full enclave layout information (only necessary ones for relocations and heap allocators) to the enclave.

haitaohuang replied:
Please see updated version with abstraction layer spec

haitaohuang commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+
+## Motivation ##
+
+An enclave's memory is backed by a special reserved region in RAM, called Enclave Page Cache (EPC). Enclave memory management tasks include allocating/reserving virtual address ranges, committing physical EPC pages, changing EPC page permissions or page type, and removing EPC pages. Those tasks require collaboration between the trusted runtime, the untrusted runtime, and the OS. The SGX enclave m ...
```
will do for next verion

mjp41 commented:
`snmalloc` can be implemented easily against the API you are providing here.  The `snmalloc` requirements are explained here:

  https://github.com/microsoft/snmalloc/blob/master/docs/PORTING.md

The only addition from `snmalloc`s perspective that might be nice is a `MAP_ALIGNED(n)` that several of the BSDs support, and the equivalent supported by `VirtualAlloc2`.  `snmalloc` will allocate large ranges, and then manage that address space to find suitable alignments, if this is not provided.  This increases `snmalloc`s address space requirements without `MAP_ALIGNED` support.

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+* @return 0 on success. EEXIST if range requested is overlapping a region in use and EMA_FIXED is requested,
+* ENOMEM for out of memory, EACCES if region is outside enclave address space.
+*/
+int sgx_mm_alloc(void *addr, size_t length, int flags, enclave_fault_handler_t handler, void* handler_private, void** out_addr)
```
```suggestion
int sgx_mm_alloc(void *address, size_t length, int flags, enclave_fault_handler_t handler, void* handler_private, void** out_address)
```
OE style guidelines say to avoid abbreviations

haitaohuang replied:
As we will position this as separate library, I'll keep abbreviations for now :-)

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+/**
+* Commit a partial or full range of memory allocated previously with EMA_COMMIT_ON_DEMAND.
+* The API will return 0 if all pages in the requested range are successfully committed. 
+* Calling this API on pages already committed has not effect. 
```
```suggestion
* Calling this API on pages already committed has no effect. 
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,350 @@
+* Calling this API on pages already committed has not effect. 
+* @param[in] addr, page aligned starting address
+* @param[in] length, length of the region in bytes of multiples of page size.
+* @return 0 on success. EINVAL if any requested page is not in any previously allocated regions, -1 otherwise 
```
Why return -1 for other failures instead of a more specific error code from https://linux.die.net/man/3/errno or from `oe_result_t`?

haitaohuang replied:
updated to return EFAULT for all other errors

dthaler commented:
What issue number is this for?   https://github.com/openenclave/openenclave/blob/master/docs/Community/Contributing.md#general-guidelines says

> * **DO** open an issue for design discussion before making any major changes.

haitaohuang commented:
> The only addition from `snmalloc`s perspective that might be nice is a `MAP_ALIGNED(n)` that several of the BSDs support, and the equivalent supported by `VirtualAlloc2`. `snmalloc` will allocate large ranges, and then manage that address space to find suitable alignments, if this is not provided. This increases `snmalloc`s address space requirements without `MAP_ALIGNED` support.

I'll add this support in next version. Thanks

haitaohuang commented:
will position EMM as a reusable library, and specify APIs of an abstraction layer for runtimes to implement. With that, coding style and naming convention of EMM would not be runtime specific.

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+address ranges, and modify attributes of the reserved/committed pages.
+
+For details of specific memory management related flows, please refer to
+[the SGX EDMM driver API spec](https://github.com/openenclave/openenclave/pull/3755/files).
```
design docs shouldn't refer to pull requests, only other docs

haitaohuang replied:
will resolve this after driver interface merged

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+**Commit, Uncommit Enclave Memory**
+
+When a page in COMMIT_ON_DEMAND allocations is accessed, a page fault occurs if
+the page was not yet commited.  The EMM will perform EACCEPT to commit the EPC
```
```suggestion
the page was not yet committed.  The EMM will perform EACCEPT to commit the EPC
```
typo

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+**Runtime Abstraction Layer**
+
+To make the EMM implementation portable across different SGX runtimes,
+e.g. OpenEnclave and Intel SGX SDKs, this document also proposes an abstraction
```
```suggestion
e.g., OpenEnclave and Intel SGX SDKs, this document also proposes an abstraction
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+ * @retval EXCEPTION_CONTINUE_EXECUTION on success handling the exception.
+ * @retval EXCEPTION_CONTINUE_SEARCH if the EMM does not handle the exception.
+ */
+typedef int (*enclave_pfhandler_t)(const sgx_pfinfo *pfinfo);
```
```suggestion
typedef int (*sgx_enclave_pfhandler_t)(const sgx_pfinfo *pfinfo);
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+/* Return value used by the EMM #PF handler to indicate
+ *  to the dispatcher that it should continue searching for the next handler.
+ */
+#define EXCEPTION_CONTINUE_SEARCH 0x0
```
```suggestion
#define SGX_EXCEPTION_CONTINUE_SEARCH 0x0
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+/* Return value used by the EMM #PF handler to indicate
+ *  to the dispatcher that it should stop searching and continue execution.
+ */
+#define EXCEPTION_CONTINUE_EXECUTION 0xFFFFFFFF
```
```suggestion
#define SGX_EXCEPTION_CONTINUE_EXECUTION 0xFFFFFFFF
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+ * portion of the buffer lies outside the enclave's memory, return false.
+ *
+ * @param[in] ptr The pointer pointer to buffer.
+ * @param[in] size The size of buffer
```
```suggestion
 * @param[in] size The size of the buffer.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+    void*               hprivate;       // private data for handler
+    ema_t*              next;           // next in doubly linked list
+    ema_t*              prev;           // prev in doubly linked list
+} ema_t;
```
```suggestion
} sgx_ema_t;
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
````diff
@@ -0,0 +1,679 @@
+
+```
+typedef BitArray;
+typedef struct _ema_t {
````
```suggestion
typedef struct _sgx_ema_t {
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
````diff
@@ -0,0 +1,679 @@
+allocated regions in enclave address space (ELRANGE).
+
+```
+typedef BitArray;
````
what is BitArray?  should it be `sgx_bit_array_t`?

haitaohuang replied:
will remove this definition. It is implementation detail and  the comments for eaccept_map should be enough for design doc.

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+ *         some other handler.
+ *
+ */
+typedef int (*enclave_fault_handler_t)(const sgx_pfinfo *pfinfo, void *private_data);
```
```suggestion
typedef int (*sgx_enclave_fault_handler_t)(const sgx_pfinfo *pfinfo, void *private_data);
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+ * @param[in] flags, a bitwise OR of flags describing committing mode, committing
+ * order, address preference, page type
+ *        Flags should include exactly one of following for committing mode:
+ *            - EMA_RESERVE: just reserve an address range, no EPC commited.
```
prefix these defines with SGX_ if they are used with a function (like this one) that starts with sgx_
```suggestion
 *            - SGX_EMA_RESERVE: just reserve an address range, no EPC committed.
```
also note typo at end :)

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+
+/*
+ * Allocate a new memory region in enclave address space (ELRANGE).
+ * @param[in] addr, starting address of the region, page aligned. If NULL is provided,
```
```suggestion
 * @param[in] addr starting address of the region, page aligned. If NULL is provided,
```
Don't put comma after param name in doxygen comments

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+ *
+ * @param[in] pfinfo, info reported in the SSA MISC region for page fault.
+ * @param[in] private_data, private data provided by handler in sgx_mm_alloc call.
+ * @retval EXCEPTION_CONTINUE_EXECUTION,  success on handling the exception.
+ * @retval EXCEPTION_CONTINUE_SEARCH, Exception not handled and should be passed
```
```suggestion
 * @retval EXCEPTION_CONTINUE_EXECUTION Success on handling the exception.
 * @retval EXCEPTION_CONTINUE_SEARCH Exception not handled and should be passed.
```

nit: Don't put comma after value name in doxygen comments, and be consistent in terms of whether the next word is capitalized or not, and whether the phrase ends in punctuation

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+ *                           The argument n specifies the binary logarithm of
+ *                           the desired alignmentand must be at least 12.
+ *        Optionally ORed with one of following page types:
+ *             - EMA_PAGE_TYPE_REG: regular page type.This is the default if not specified.
```
```suggestion
 *             - EMA_PAGE_TYPE_REG: regular page type. This is the default if not specified.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+ *             - EMA_PAGE_TYPE_SS_REST: the rest page in shadow stack. 
+ * @retval 0 on success.
+ * @retval EINVAL if any parameter passed in is not valid.
+ * @retval errno reported by dependent syscalls, e.g. mmap().
```
```suggestion
 * @retval errno reported by dependent syscalls, e.g., mmap().
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,679 @@
+ *             - EMA_PAGE_TYPE_REG: regular page type.This is the default if not specified.
+ *             - EMA_PAGE_TYPE_SS_FIRST: the first page in shadow stack.
+ *             - EMA_PAGE_TYPE_SS_REST: the rest page in shadow stack. 
+ * @retval 0 on success.
```
```suggestion
 * @retval 0 The operation was successful.
```
Doxygen generates a table, so after the return value should be a statement to appear in the 2nd column of the table.  Recommend phrasing it like an error message, as a sentence.

anakrish commented:
@haitaohuang Thanks for this proposal. Looks very good.
My request would be to see if we can support SGX1 too using the same Memory Manager interface. That way OE could implement a mmap/munmap that just forwards to the Memory Manager and everything would automatically work without code changes when moving between SGX1 and SGX2. It is Ok that only a subset of the functionality works on SGX1 due to hardware limitation.

For reference, here is Mystikos' mmap implementation based on SGX1.
https://github.com/deislabs/mystikos/blob/main/kernel/mman.c

haitaohuang commented:
@anakrish IIUC Mystikos mman implementation can still use the APIs here to reserve the whole range [base, end) with mm_alloc(..., SGX_EMA_COMMIT_ON_DEMAND), then reuse the same code to manage its BRK, MAP and VADs to track regions it allocates for user. The other differences on SGX2 now are following:
1. It needs to do sgx_mm_commit for each new allocation (triggered by sbrk, mmap, mremap).
2. sgx_mm_uncommit for deallocation (triggered by sbrk, munmap, mremap).
3. sgx_mm_modify_permission to change permissions in secure way.

For implementation on SGX1, it does not need to make any of those calls because the memory region is allocated and committed and no EPCM permissions change possible. So I expect those SGX2 code flow can be #ifdef out for SGX1 compilation.

On the other hand, for EMM to support both SGX1 and SGX2, we need add these limitations for SGX1:
- sgx_mm_allocate treats COMMIT_NOW and COMMIT_ON_DEMAND the same on SGX1.
- sgx_mm_commit/sgx_mm_uncommit has no effect.
- sgx_mm_commit_data on SGX1 platform will fail with an error code.
- sgx_mm_dealloc on SGX1 does not trim EPC and only marks pages as free forfuture allocations.
- sgx_mm_modify_page_type is not supported.
- sgx_mm_modify_ex/sgx_mm_modify_permissions will attempt to change OS PTE permissions only,
and can not enforce permissions in a trusted way on SGX1 platform.

It might save a few lines of code for upper layer by avoiding some #ifdefs. But I'm not sure it worth adding this extra layer for SGX1 runtime though.  
Your thoughts?

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+
+**Allocate, Deallocate Enclave Memory**
+
+The EMM provides an API, sgx_mm_alloc, for clients to request enclave memory
```
Clarify: when this says "the EMM provides an API", which SDK/library is this referring to?  PSW?  The Intel SGX SDK for Linux?  The Intel SGX SDK for Windows?  (Presumably it's not the OE SDK or the API name would start with oe_ instead.)  And what is the "client" here, is it an app or the OE SDK?  (I *think* this is answered down at line 78 but since it's way below, this text is confusing.  I might suggest moving the Runtime Abstraction Layer section starting at 72 up before this section.)

haitaohuang replied:
will move up the abstraction layer section to give more context. thx

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+**Runtime Abstraction Layer**
+
+To make the EMM implementation portable across different SGX runtimes,
+e.g., OpenEnclave and Intel SGX SDKs, this document also proposes an abstraction
```
```suggestion
e.g., Open Enclave and Intel SGX SDKs, this document also proposes an abstraction
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+
+The enclave memory manager keeps track of memory allocation and layout info inside
+enclave address range (ELRANGE) using an internal structure called the Enclave Memory
+Area (EMA) List. The EMA and EMA list are considered private data structure of memory
```
```suggestion
Area (EMA) List. The EMA and the EMA list are considered private data structures of the memory
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+
+**Assumptions:**
+
+- When enclave is loaded, the OS reserves the whole address range covered by ELRANGE.
```
```suggestion
- When an enclave is loaded, the OS reserves the whole address range covered by ELRANGE.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+
+- When enclave is loaded, the OS reserves the whole address range covered by ELRANGE.
+It is assumed the host app will not remap any part of this reserved range.
+- When an enclave is loaded with base address at zero, only partial ELRANGE may be
```
```suggestion
- When an enclave is loaded with base address at zero, only a partial ELRANGE may be
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+- When enclave is loaded, the OS reserves the whole address range covered by ELRANGE.
+It is assumed the host app will not remap any part of this reserved range.
+- When an enclave is loaded with base address at zero, only partial ELRANGE may be
+  reserved by the OS. In that case, the EMM will assume the partial ELRANGE as valid reserved
```
```suggestion
  reserved by the OS. In that case, the EMM will assume the partial ELRANGE as a valid reserved
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+- When an enclave is loaded with base address at zero, only partial ELRANGE may be
+  reserved by the OS. In that case, the EMM will assume the partial ELRANGE as valid reserved
+  range for use inside the enclave.
+- Memory manager does not check EPC pressure, or proactively trim pages when EPC runs low.
```
```suggestion
- The memory manager does not check EPC pressure, or proactively trim pages when EPC runs low.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+- Memory manager does not check EPC pressure, or proactively trim pages when EPC runs low.
+The OS can reclaim EPC pages when EPC running low or cgroups threshold reached
+- Memory manager does not maintain and recycle committed then freed pages
+  - Whenever a page is freed (via dealloc or uncommit API), it is trimmed from enclave
+  and need be re-allocated and committed before re-use.
```
```suggestion
  - Whenever a page is freed (via dealloc or uncommit API), it is trimmed from the enclave
  and needs to be re-allocated and committed before re-use.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+- Memory manager does not maintain and recycle committed then freed pages
+  - Whenever a page is freed (via dealloc or uncommit API), it is trimmed from enclave
+  and need be re-allocated and committed before re-use.
+  - Owner of a region can re-purpose a sub-region of it by calling sgx_mm_modify_type/permissions
```
```suggestion
  - The owner of a region can re-purpose a sub-region of it by calling sgx_mm_modify_type/permissions
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+  and need be re-allocated and committed before re-use.
+  - Owner of a region can re-purpose a sub-region of it by calling sgx_mm_modify_type/permissions
+  to split out the sub-region to be reused.
+- Memory manager does not call back client for #GP handling. Memory manager code will ensure
```
```suggestion
- The memory manager does not call back into the client for #GP handling. Memory manager code will ensure that
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+  to split out the sub-region to be reused.
+- Memory manager does not call back client for #GP handling. Memory manager code will ensure
+itself would not cause #GP, and only register a #PF handler with the enclave global exception
+handler registry through the runtime abstraction layer. Clients wish to handle #GP can register
```
```suggestion
handler registry through the runtime abstraction layer. A client wishing to handle #GP can register
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+ *
+ * @param[in] pfinfo info reported in the SSA MISC region for page fault.
+ * @param[in] private_data private data provided by handler in sgx_mm_alloc call.
+ * @retval SGX_EXCEPTION_CONTINUE_EXECUTION  Success on handling the exception.
```
```suggestion
 * @retval SGX_EXCEPTION_CONTINUE_EXECUTION Success on handling the exception.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+ *                  then the function will select the starting address.
+ * @param[in] length size of the region in multiples of page size in bytes.
+ * @param[in] flags a bitwise OR of flags describing committing mode, committing
+ * order, address preference, page type
```
```suggestion
 * order, address preference, and page type.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+**Remarks:**
+- Permissions of newly allocated regions are always SGX_EMA_PROT_READ|SGX_EMA_PROT_WRITE and of page
+    type SGX_EMA_PAGE_TYPE_REG, except for SGX_EMA_RESERVE mode regions which will have SGX_EMA_PROT_NONE.
+- Once allocated by sgx_mm_alloc, a region will stay in allocated state and become
```
```suggestion
- Once allocated by sgx_mm_alloc, a region will stay in the allocated state and become
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+ * Deallocate the address range.
+ * The pages in the allocation are freed and the address range is released for future allocation.
+ * @param[in] addr page aligned start address of the region to be freed and released
+ * @param[in] length size in bytes of multiples of page size.
```
```suggestion
 * @param[in] length size in bytes in multiples of page size.
```

dthaler replied:
Some places in this doc have "of multiples" and others say "in multiples".  I am not sure if they are intended to be the same.  Does this mean that the length must be a multiple of page size, as opposed to being the number of multiples of page size?

haitaohuang replied:
length should be in bytes but the number value should be multiples of page size, k*4096. So is `of multiples of page size` better?

dthaler replied:
```suggestion
 * @param[in] length size in bytes.  The length must be a multiple of the page size.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+/*
+ * Deallocate the address range.
+ * The pages in the allocation are freed and the address range is released for future allocation.
+ * @param[in] addr page aligned start address of the region to be freed and released
```
```suggestion
 * @param[in] addr page aligned start address of the region to be freed and released.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
````diff
@@ -0,0 +1,678 @@
+
+```
+/*
+ * Change permissions and/or page type of a previously allocated region
````
```suggestion
 * Change permissions and/or page type of a previously allocated region.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
````diff
@@ -0,0 +1,678 @@
+```
+**Remarks:**
+- The memory manager will track current permissions for each region, and can
+    determine whether new permissions require an OCall for EMODPR, e.g RW<->RX, RW->R.
````
```suggestion
    determine whether new permissions require an OCall for EMODPR, e.g., RW<->RX, RW->R.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
````diff
@@ -0,0 +1,678 @@
+
+```
+**Remarks:**
+- The memory manager decides whether OCalls are needed to request OS to make PTE
````
```suggestion
- The memory manager decides whether OCalls are needed to ask the OS to make Page Table Entry (PTE)
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+/*
+ * The EMM page fault (#PF) handler
+ *
+ * @param[in] pfinfo info reported in the SSA MISC region for page fault
```
```suggestion
 * @param[in] pfinfo info reported in the SSA MISC region for page fault.
```
nit: Be consistent throughout this file in terms of whether param descriptions end in a period or not.

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+bool sgx_register_pfhandler(sgx_enclave_pfhandler_t pfhandler);
+
+/**
+ * Unregister the EMM handler with the global exception handler registry
```
```suggestion
 * Unregister the EMM handler with the global exception handler registry.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
````diff
@@ -0,0 +1,678 @@
+
+```
+/*
+ * Call OS to reserve region for EAUG, immediately or on-demand
````
```suggestion
 * Call OS to reserve region for EAUG, immediately or on-demand.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+ *
+ * @param[in] addr desired page aligned start address, NULL if no desired address
+ * @param[in] length size of the region in multiples of page size in bytes
+ * @param[in] flags  a bitwise OR of flags describing committing mode, committing
```
```suggestion
 * @param[in] flags a bitwise OR of flags describing committing mode, committing
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+ * @param[in] addr desired page aligned start address, NULL if no desired address
+ * @param[in] length size of the region in multiples of page size in bytes
+ * @param[in] flags  a bitwise OR of flags describing committing mode, committing
+ *                     order, address preference, page type. The untrusted side
+ *    implementation should always invoke mmap syscall with MAP_SHARED|MAP_FIXED_NOREPLACE, and
```
nit: shouldn't these two lines have the same indentation?

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+int sgx_mmap_ocall(void *addr, size_t length, int prot, int flags);
+
+/*
+ * Call OS to change permissions, type, or notify EACCEPT done after TRIM
```
```suggestion
 * Call OS to change permissions, type, or notify EACCEPT done after TRIM.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+ *            SGX_EMA_PAGE_TYPE_TCS: change the page type to PT_TCS
+ *            SGX_EMA_PROT_NOACCESS: Signal the kernel EACCEPT is done for PT_TRIM pages.
+ * @retval 0 The operation was successful.
+ * @retval EINVAL if any parameter passed in is not valid.
```
```suggestion
 * @retval EINVAL A parameter passed in is not valid.
```
retval generates a table where everything after the first token is in a separate column.  Use complete sentences line lines 500 and 502 do, no "if"

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+In addition to implement the abstraction layer APIs, a runtime shall provide
+iniitial enclave memory layout information to the EMM during early
+initialization phase of the enclave.
+The memory manager must be initialized in first ECALL (ECMD_INIT_ENCLAVE in
```
```suggestion
The memory manager must be initialized in the first ECALL (ECMD_INIT_ENCLAVE in
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+iniitial enclave memory layout information to the EMM during early
+initialization phase of the enclave.
+The memory manager must be initialized in first ECALL (ECMD_INIT_ENCLAVE in
+Intel SGX SDK) before any other clients can use it. Therefore, code and data
```
When this document says "Intel SGX SDK" does it refer to both the Intel SGX SDK for Linux and the Intel SGX SDK for Windows, since those are different SDKs?  Clarify.  Avoid ever using the term "Intel SGX SDK" without "for ..." since the term is ambiguous/confusing.

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+with EADD before EINIT, and as a part of the trusted runtime.
+
+The trusted runtime should enumerate all initial committed regions (code,
+data, heap, stack, TCS, SSA), and call the EMM internal APIs to setup
```
```suggestion
data, heap, stack, TCS, and SSA), and call the EMM internal APIs to set up
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+The trusted runtime should enumerate all initial committed regions (code,
+data, heap, stack, TCS, SSA), and call the EMM internal APIs to setup
+initial entries in the EMA list to track existing regions and mark some
+of them not modifiable by EMM public APIs. Runtime also ensures there is
```
```suggestion
of them as not modifiable by EMM public APIs. The runtime also ensures there is
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+data, heap, stack, TCS, SSA), and call the EMM internal APIs to setup
+initial entries in the EMA list to track existing regions and mark some
+of them not modifiable by EMM public APIs. Runtime also ensures there is
+enough reserved space on heap for EMM to create the initial EMA list and
```
```suggestion
enough reserved space on the heap for EMM to create the initial EMA list and
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+below certain threshold.
+
+Alternative option: At build time, the enclave signing tool can precalculate
+and fill in EMA entries that holds info of initial regions to be committed by
```
```suggestion
and fill in EMA entries that hold info on initial regions to be committed by
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+Alternative option: At build time, the enclave signing tool can precalculate
+and fill in EMA entries that holds info of initial regions to be committed by
+EADD during enclave load. The calculated start addresses in these EMAs can be
+relative to enclave secs->base. Runtime can patch those entries at
```
```suggestion
relative to enclave secs->base. The runtime can patch those entries at
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+Internal APIs and Structures
+-------------------------------------
+
+Following are internal functions and structures to be used by EMM implementation.
```
```suggestion
The following are internal functions and structures to be used by the EMM implementation.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+-------------------------------------
+
+Following are internal functions and structures to be used by EMM implementation.
+They can evolve over time, shown here for reference only.
```
```suggestion
They can evolve over time, and are shown here for reference only.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
````diff
@@ -0,0 +1,678 @@
+
+```
+ **Remarks:**
+ - Access to the list (find, insert, remove EMAs) are synchronized for thread-safety.
````
```suggestion
 - Accesses to the list (find, insert, remove EMAs) are synchronized for thread-safety.
```

dthaler commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+Metadata,  File format
+---------------------------------------
+
+Enclave metadata and file format are runtime specific. Detailed design is
```
```suggestion
The enclave metadata and file format are runtime specific. A detailed design is
```

mingweishih commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+## Notes on Internal Design ##
+
+The enclave memory manager keeps track of memory allocation and layout info inside
+enclave address range (ELRANGE) using an internal structure called the Enclave Memory
```
Does this mean the enclave need to over request the ELRANGE during the creation?
Something like (where `ADDITIONAL REGION` is for the enclave to dynamically allocate via EMM):
```
[ENCLAVE LAYOUT][ADDITIONAL REGION]
<----          ELRANGE        ---->
```

And when 0-base enclave is enabled, the layout becomes

```
[REGION from ADDR 0][ENCLAVE LAYOUT][ADDITIONAL REGION]
<----                    ELRANGE                  ---->
```
I assume the EMM internally will ensure that `REGION from ADDR 0` should not be used?

haitaohuang replied:
Runtime can return false from sgx_is_within_enclave or ema_alloc a RESERVED region for REGION from ADDR 0 with EMA_SYSTEM flag so no one else can allocate from it.

haitaohuang commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,678 @@
+
+### Support for EMM Initialization
+
+In addition to implement the abstraction layer APIs, a runtime shall provide
```
The initial idea was for runtime to call ema_alloc with EMA_SYSTEM flag to set up those initial layouts. And other private APIs below can be used to adjust attributes and boundaries of initially committed pages

Now I am also thinking to add one more API specific for EADDed regions, some thing like this:
```
/*
 * Initialize an EMA. This can be used to setup EMAs to account regions that
 * are loaded and initialized with EADD before EINIT.  
 * @param[in] addr Starting address of the region, page aligned. If NULL is provided,
 *                  then the function will select the starting address.
 * @param[in] length Size of the region in multiples of page size in bytes.
 * @param[in] flags SGX_EMA_SYSTEM, or SGX_EMA_SYSTEM | SGX_EMA_RESERVE
 *           bitwise ORed with one of following page types:
 *             - SGX_EMA_PAGE_TYPE_REG: regular page type. This is the default if not specified.
 *             - SGX_EMA_PAGE_TYPE_TCS: TCS page.
 *             - SGX_EMA_PAGE_TYPE_SS_FIRST: the first page in shadow stack.
 *             - SGX_EMA_PAGE_TYPE_SS_REST: the rest page in shadow stack.
 * @param[in] prot permissions, either SGX_EMA_PROT_NONE or a bitwise OR of following with:
 *        - SGX_EMA_PROT_READ: Pages may be read.
 *        - SGX_EMA_PROT_WRITE: Pages may be written.
 *        - SGX_EMA_PROT_EXECUTE: Pages may be executed.
 * @param[in] handler A custom handler for page faults in this region, NULL if
 *                     no custom handling needed.
 * @param[in] handler_private Private data for the @handler, which will be passed
 *                     back when the handler is called.
 * @retval 0 The operation was successful.
 * @retval EACCES Region is outside enclave address space.
 * @retval EEXIST Any page in range requested is in use.
 * @retval EINVAL Invalid page type, flags, or addr and length are not page aligned.
 */ 
int ema_init(void *addr, size_t length, int flags, int prot,
              sgx_enclave_fault_handler_t handler,
              void *handler_private);
```

dthaler reviewed (approved):
Understood that this document might move to an Intel repo and just be referenced from an OE document, since this is info about SGX not about the OE SDK per se.

mingweishih commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,712 @@
+ * Define a mutex and create/lock/unlock/destroy functions.
+ */
+typedef struct _sgx_mm_mutex sgx_mm_mutex;
+sgx_mm_mutex *sgx_mm_mutex_create(void);
```
Does this assume creating a non-recursive mutex?

haitaohuang commented on docs/DesignDocs/SGX_EMM.md:
```diff
@@ -0,0 +1,712 @@
+ * @param[in] user_start The start of the user address range, page aligned.
+ * @param[in] user_end The end (exclusive) of the user address range, page aligned.
+ */
+void sgx_mm_init(size_t user_start, size_t user_end);
```
This function only try to reserve an initial area (16 pages) for internal use and will call abort() if fails for some reason, e.g. there is not enough in [user_start, user_end).  Was the hang on this API or subsequent calls to other APIs?

haitaohuang replied:
@mingweishih I will change to return ENOMEM if internal allocation failed for initial area. Let me know if more concerns.

mingweishih replied:
Sounds good to me. Can you update this document as well?

haitaohuang replied:
updated now.

haitaohuang commented:
Updated doc to reflect changes for ocalls to align with the new functions specified in Enclave Common, which are used for OCall implementation.
Also added note for runtime implemented mutex to be recursive.
Note as discussed in OE arch meeting, we moved EMM code to separate SGX EMM repo: https://github.com/intel/sgx-emm
Though review to continue on this PR and I will keep sync changes made in the EMM repo to this PR for review purpose, we won't need to merge this PR into OE.

haitaohuang commented:
Close without merge as commented earlier.
If any concerns in future, please open issues in https://github.com/intel/sgx-emm
